modules/apps/cookies/cookies-impl/src/main/java/com/liferay/cookies/internal/events/CookiesPreAction.java

```diff
@@ -0,0 +1,225 @@
+
+					cookieMap.put(cookieName, cookie.getValue());
+				}
+			}
+		}
+
+		return cookieMap;
+	}
+
+	private void _run(
+			HttpServletRequest httpServletRequest,
+			HttpServletResponse httpServletResponse)
+		throws ConfigurationException {
+
+		ThemeDisplay themeDisplay =
+			(ThemeDisplay)httpServletRequest.getAttribute(
+				WebKeys.THEME_DISPLAY);
+
+		CookiesPreferenceHandlingConfiguration
+			cookiesPreferenceHandlingConfiguration =
+				_configurationProvider.getGroupConfiguration(
+					CookiesPreferenceHandlingConfiguration.class,
+					themeDisplay.getScopeGroupId());
+
+		Map<String, String> requestCookies = _parseCookieMap(
+			httpServletRequest.getCookies());
+
+		if (!cookiesPreferenceHandlingConfiguration.enabled()) {
+			_expireCookiesIfSet(
+				httpServletRequest, httpServletResponse, requestCookies,
+				CookiesConstants.NAME_USER_CONSENT_CONFIGURED,
+				CookiesConstants.NAME_CONSENT_TYPE_NECESSARY,
+				CookiesConstants.NAME_CONSENT_TYPE_PERFORMANCE,
+				CookiesConstants.NAME_CONSENT_TYPE_PERSONALIZATION,
+				CookiesConstants.NAME_CONSENT_TYPE_FUNCTIONAL);
+		}
+		else {
+			boolean necessaryConsentCookieSet = Validator.isNotNull(
+				requestCookies.get(
+					CookiesConstants.NAME_CONSENT_TYPE_NECESSARY));
+			boolean performanceConsentCookieSet = Validator.isNotNull(
+				requestCookies.get(
+					CookiesConstants.NAME_CONSENT_TYPE_PERFORMANCE));
+			boolean functionalConsentCookieSet = Validator.isNotNull(
+				requestCookies.get(
+					CookiesConstants.NAME_CONSENT_TYPE_FUNCTIONAL));
+			boolean personalizationConsentCookieSet = Validator.isNotNull(
+				requestCookies.get(
+					CookiesConstants.NAME_CONSENT_TYPE_PERSONALIZATION));
+
+			boolean optionalConsentCookiesSet = false;
+
+			if (performanceConsentCookieSet && functionalConsentCookieSet &&
+				personalizationConsentCookieSet) {
+
+				optionalConsentCookiesSet = true;
+			}
+
+			boolean userConsentCookieSet = Validator.isNotNull(
+				requestCookies.get(
+					CookiesConstants.NAME_USER_CONSENT_CONFIGURED));
+
+			if (!(optionalConsentCookiesSet && necessaryConsentCookieSet &&
+				  userConsentCookieSet)) {
+
+				if (!cookiesPreferenceHandlingConfiguration.
+						explicitConsentMode() ||
+					!userConsentCookieSet) {
+
+					_expireCookiesIfSet(
+						httpServletRequest, httpServletResponse, requestCookies,
+						CookiesConstants.NAME_USER_CONSENT_CONFIGURED);
+				}
+
+				if (!necessaryConsentCookieSet) {
+					_addCookies(
+						httpServletRequest, httpServletResponse,
+						new Cookie(
+							CookiesConstants.NAME_CONSENT_TYPE_NECESSARY,
+							"true"));
+				}
+
+				if (!optionalConsentCookiesSet ||
+					(cookiesPreferenceHandlingConfiguration.
+						explicitConsentMode() &&
+					 !userConsentCookieSet)) {
+
+					String optionalCookiesValue = "true";
+
+					if (cookiesPreferenceHandlingConfiguration.
+							explicitConsentMode()) {
+
+						optionalCookiesValue = "false";
+					}
```

Comment:
I have to say that having to do this sort of stuff is soul crushing, this would be 1-2 lines with a ternary operator 🥲